File: tools/fleet-customer-metrics/import_data.py
import os
import csv
import io
import sys
import json
import re
import requests
import psycopg2

def main():
    """
    Main function to download, cleanse, and import CSV data into PostgreSQL.
    """
    # --- Configuration ---
    try:
        db_host = os.environ['PGHOST']
        db_port = os.environ['PGPORT']
        db_name = os.environ['PGDATABASE']
        db_user = os.environ['PGUSER']
        db_password = os.environ['PGPASSWORD']
        csv_url = os.environ['CSV_URL']
    except KeyError as e:
        print(f"Error: Environment variable {e} not set.")
        sys.exit(1)

    db_table = "usage_data"
    exclude_column = 'id'

    integer_columns = [
        'numHostsEnrolled', 'numUsers', 'numTeams', 'numQueries', 'numPolicies',
        'numLabels', 'numWeeklyActiveUsers', 'numHostsNotResponding',
        'numHostsABMPending', 'numSoftwareVersions', 'numHostSoftwares', 'numSoftwareTitles',
        'numHostSoftwareInstalledPaths', 'numSoftwareCPEs', 'numSoftwareCVEs', 'numHostsFleetDesktopEnabled'
    ]
    json_columns = [
        'hostsEnrolledByOperatingSystem', 'hostsEnrolledByOrbitVersion', 'hostsEnrolledByOsqueryVersion'
    ]

    # --- 1. Download the CSV data ---
    print(f"Downloading CSV data from {csv_url}...")
    try:
        response = requests.get(csv_url, timeout=30)
        response.raise_for_status()
        print("Download complete.")
    except requests.exceptions.RequestException as e:
        print(f"Error downloading CSV: {e}")
        sys.exit(1)

    # --- 2. Cleanse the CSV data in memory ---
    print("Cleansing data in memory...")

    # Use a regex that can split by comma but ignore commas inside quotes
    # This is necessary because the source CSV is not properly formatted
    csv_regex = re.compile(r',(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)')

    # Read the data line by line
    lines = response.text.strip().split('\n')

    # Manually parse the header
    header = [h.strip('"') for h in csv_regex.split(lines[0])]

    # Prepare the in-memory files for cleaned data
    cleaned_data_file = io.StringIO()
    writer = csv.writer(cleaned_data_file, lineterminator='\n')

    try:
        exclude_index = header.index(exclude_column)
        time_index = header.index('updatedTime')
        integer_indices = {col: header.index(col) for col in integer_columns if col in header}
        json_indices = {col: header.index(col) for col in json_columns if col in header}
    except ValueError as e:
        print(f"Error: Column {e} not found in the CSV header.")
        sys.exit(1)

    new_header = [col for i, col in enumerate(header) if i != exclude_index]
    writer.writerow(new_header)

    for line in lines[1:]: # Start from the second line (skip header)
        row = [field.strip('"') for field in csv_regex.split(line)]
        # This check is important for rows that might have a different number of columns
        if len(row) != len(header):
            print(f"Warning: Skipping malformed row with {len(row)} columns: {line[:100]}...")
            continue

        # Clean datetime field
        if time_index != -1 and '+00' in row[time_index]:
            row[time_index] = row[time_index].replace('+00', '')

        # Clean integer fields
        for col_name, index in integer_indices.items():
            if row[index] and '.' in row[index]:
                try:
                    row[index] = str(int(float(row[index])))
                except (ValueError, TypeError):
                    print(f"Warning: Could not convert '{row[index]}' to int for column '{col_name}'.")

        # Clean and re-encode JSON fields
        for col_name, index in json_indices.items():
            if row[index]:
                try:
                    # Strip outer quotes, then replace the double-escaped quotes
                    json_str = row[index].strip('"').replace('""', '"')
                    # Parse the now-valid JSON string
                    json_obj = json.loads(json_str)
                    # Dump it back into a compact, valid JSON string for the database
                    row[index] = json.dumps(json_obj, separators=(',', ':'))
                except json.JSONDecodeError:
                    print(f"Warning: Could not parse JSON for column '{col_name}'. Setting to null.")
                    row[index] = None # Fallback in case of other unexpected errors

        new_row = [field for i, field in enumerate(row) if i != exclude_index]
        writer.writerow(new_row)

    cleaned_data_file.seek(0)
    print("Cleansing complete.")


    # --- 3. Import the cleansed CSV into PostgreSQL ---
    conn = None
    try:
        print(f"Connecting to database '{db_name}'...")
        conn = psycopg2.connect(
            dbname=db_name,
            user=db_user,
            password=db_password,
            host=db_host,
            port=db_port
        )
        print("Connection successful.")

        with conn.cursor() as cursor:
            # Ensure destination table exists with reasonable column types inferred from header
            print(f"Ensuring table '{db_table}' exists...")

            ddl_columns = []
            for col in new_header:
                if col == 'updatedTime':
                    col_type = 'TIMESTAMP'
                elif col in integer_columns:
                    col_type = 'INTEGER'
                elif col in json_columns:
                    col_type = 'JSONB'
                else:
                    col_type = 'TEXT'
                ddl_columns.append(f'"{col}" {col_type}')

            create_table_sql = f"CREATE TABLE IF NOT EXISTS {db_table} (" + ", ".join(ddl_columns) + ");"
            cursor.execute(create_table_sql)
            conn.commit()

            print(f"Importing data into table '{db_table}'...")

            quoted_header = ','.join([f'"{col}"' for col in new_header])

            cursor.copy_expert(
                sql=f"COPY {db_table} ({quoted_header}) FROM STDIN WITH (FORMAT CSV, HEADER)",
                file=cleaned_data_file
            )
            conn.commit()
            print(f"{cursor.rowcount} rows imported successfully.")

    except psycopg2.Error as e:
        print(f"Database error: {e}")
        if conn:
            conn.rollback()
        sys.exit(1)
    finally:
        if conn:
            conn.close()
            print("Database connection closed.")

if __name__ == "__main__":
    main()
    print("Process finished.")

File: tools/fleet-customer-metrics/import_data.ts
import { Client } from "pg";
import { from as copyFrom } from "pg-copy-streams";
import { Readable } from "stream";
import { pipeline } from "stream/promises";

const DB_TABLE = "usage_data";

const INTEGER_COLUMNS = new Set([
  "numHostsEnrolled",
  "numUsers",
  "numTeams",
  "numQueries",
  "numPolicies",
  "numLabels",
  "numWeeklyActiveUsers",
  "numHostsNotResponding",
  "numHostsABMPending",
  "numSoftwareVersions",
  "numHostSoftwares",
  "numSoftwareTitles",
  "numHostSoftwareInstalledPaths",
  "numSoftwareCPEs",
  "numSoftwareCVEs",
  "numHostsFleetDesktopEnabled",
]);

const JSON_COLUMNS = new Set([
  "hostsEnrolledByOperatingSystem",
  "hostsEnrolledByOrbitVersion",
  "hostsEnrolledByOsqueryVersion",
]);

function colTypeFor(col: string): string {
  if (col === "updatedTime") return "TIMESTAMP";
  if (INTEGER_COLUMNS.has(col)) return "INTEGER";
  if (JSON_COLUMNS.has(col)) return "JSONB";
  if (
    col.endsWith("Enabled") ||
    col.endsWith("Disabled") ||
    col.endsWith("Configured")
  )
    return "BOOLEAN";
  return "TEXT";
}

function csvEscape(field: string): string {
  if (field.includes(",") || field.includes('"') || field.includes("\n")) {
    return '"' + field.replace(/"/g, '""') + '"';
  }
  return field;
}

interface DataclipJSON {
  fields: string[];
  values: unknown[][];
}

async function loadData(): Promise<DataclipJSON> {
  const dataclipUrl = process.env.DATACLIP_URL;
  if (!dataclipUrl) {
    throw new Error("DATACLIP_URL environment variable is required");
  }
  const herokuToken = process.env.HEROKU_API_TOKEN;
  if (!herokuToken) {
    throw new Error("HEROKU_API_TOKEN environment variable is required");
  }
  console.log(`Fetching data from dataclip URL...`);
  const res = await fetch(dataclipUrl, {
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${herokuToken}`,
    },
  });
  if (!res.ok) {
    throw new Error(`Dataclip fetch failed: ${res.status} ${res.statusText}`);
  }
  return (await res.json()) as DataclipJSON;
}

async function main() {
  // pg Client reads PGHOST, PGPORT, PGDATABASE, PGUSER, PGPASSWORD automatically

  // --- 1. Load JSON data ---
  let data: DataclipJSON;
  try {
    data = await loadData();
  } catch (err) {
    console.error(`Error loading data: ${err}`);
    process.exit(1);
  }

  const header = data.fields;
  const dataRows = data.values;
  console.log(`Loaded ${dataRows.length} rows with ${header.length} columns.`);

  // --- 2. Cleanse data and build CSV ---
  console.log("Cleansing data...");

  const timeIndex = header.indexOf("updatedTime");
  const integerIndices = new Map<string, number>();
  for (const col of INTEGER_COLUMNS) {
    const idx = header.indexOf(col);
    if (idx !== -1) integerIndices.set(col, idx);
  }
  const jsonIndices = new Map<string, number>();
  for (const col of JSON_COLUMNS) {
    const idx = header.indexOf(col);
    if (idx !== -1) jsonIndices.set(col, idx);
  }

  const csvLines: string[] = [header.map(csvEscape).join(",")];

  for (const row of dataRows) {
    if (row.length !== header.length) {
      console.log(`Warning: Skipping malformed row with ${row.length} columns.`);
      continue;
    }

    const csvRow: string[] = row.map((val, i) => {
      if (val === null || val === undefined) return "";

      // Clean datetime field
      if (i === timeIndex && typeof val === "string" && val.includes("+00")) {
        return val.replace("+00", "");
      }

      // Clean integer fields (floats like 14.0 -> 14)
      if (integerIndices.has(header[i])) {
        const num = Number(val);
        if (!isNaN(num)) return Math.trunc(num).toString();
        console.log(`Warning: Could not convert '${val}' to int for column '${header[i]}'.`);
        return String(val);
      }

      // Serialize JSON fields
      if (jsonIndices.has(header[i]) && typeof val === "object") {
        return JSON.stringify(val);
      }

      return String(val);
    });

    csvLines.push(csvRow.map(csvEscape).join(","));
  }

  const cleanedCsv = csvLines.join("\n") + "\n";
  console.log("Cleansing complete.");

  // --- 3. Import into PostgreSQL ---
  const client = new Client();
  try {
    console.log("Connecting to database...");
    await client.connect();
    console.log("Connection successful.");

    // Ensure table exists
    console.log(`Ensuring table '${DB_TABLE}' exists...`);
    const ddlColumns = header
      .map((col) => `"${col}" ${colTypeFor(col)}`)
      .join(", ");
    await client.query(
      `CREATE TABLE IF NOT EXISTS ${DB_TABLE} (${ddlColumns});`
    );

    // Add any new columns that don't yet exist in the table
    const { rows } = await client.query(
      "SELECT column_name FROM information_schema.columns WHERE table_name = $1",
      [DB_TABLE]
    );
    const existingColumns = new Set(
      rows.map((r: { column_name: string }) => r.column_name)
    );
    for (const col of header) {
      if (!existingColumns.has(col)) {
        const colType = colTypeFor(col);
        console.log(
          `Adding new column '${col}' (${colType}) to table '${DB_TABLE}'...`
        );
        await client.query(
          `ALTER TABLE ${DB_TABLE} ADD COLUMN "${col}" ${colType};`
        );
      }
    }

    // Clear existing data before import to avoid duplicates across runs
    await client.query(`TRUNCATE ${DB_TABLE};`);

    // Bulk import via COPY
    console.log(`Importing data into table '${DB_TABLE}'...`);
    const quotedHeader = header.map((col) => `"${col}"`).join(",");
    const copyStream = client.query(
      copyFrom(
        `COPY ${DB_TABLE} (${quotedHeader}) FROM STDIN WITH (FORMAT CSV, HEADER)`
      )
    );
    await pipeline(Readable.from([cleanedCsv]), copyStream);
    console.log(`${copyStream.rowCount} rows imported successfully.`);

    // Create customers view used by Grafana dashboards.
    // Organizations with multiple environments get labeled by host count:
    // the largest is "(Prod)", a single smaller one is "(Stage)", and
    // additional ones are numbered "(Stage 2)", "(Stage 3)", etc.
    await client.query(`
      CREATE OR REPLACE VIEW customers AS
      WITH latest AS (
        SELECT "anonymousIdentifier", organization,
          MAX("numHostsEnrolled") AS max_hosts
        FROM ${DB_TABLE}
        GROUP BY "anonymousIdentifier", organization
      ),
      ranked AS (
        SELECT *,
          ROW_NUMBER() OVER (PARTITION BY organization ORDER BY max_hosts DESC) AS env_rank,
          COUNT(*) OVER (PARTITION BY organization) AS env_count
        FROM latest
      )
      SELECT "anonymousIdentifier",
        CASE
          WHEN env_count = 1 THEN organization
          WHEN env_rank = 1 THEN organization || ' (Prod)'
          WHEN env_count = 2 THEN organization || ' (Stage)'
          ELSE organization || ' (Stage ' || (env_rank - 1) || ')'
        END AS organization
      FROM ranked;
    `);
    console.log("Created customers view.");
  } catch (err) {
    console.error(`Database error: ${err}`);
    process.exit(1);
  } finally {
    await client.end();
    console.log("Database connection closed.");
  }
}

main().then(() => console.log("Process finished."));
